### ISSUE TITLE
[Phase 2] Score Network Architecture

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/diffusion-boltzmann-sampler#4]
## Objective

Implement the neural network that learns the score function for the diffusion model.

## Deliverables

- [ ] `ScoreNetwork` with time conditioning (U-Net style)
- [ ] `SinusoidalTimeEmbedding` for diffusion time
- [ ] `DiffusionProcess` with forward noising
- [ ] Tests verifying output shapes and gradient flow

## Files to Create

| File | Description | Lines |
|------|-------------|-------|
| `backend/ml/models/score_network.py` | U-Net style score network | ~150 |
| `backend/ml/models/diffusion.py` | Forward/reverse processes | ~100 |
| `backend/ml/models/noise_schedule.py` | Beta schedules | ~50 |
| `backend/tests/test_models.py` | Model tests | ~80 |

## Core Implementation

```python
# backend/ml/models/score_network.py
import torch
import torch.nn as nn
import math

class SinusoidalTimeEmbedding(nn.Module):
    """Sinusoidal position embeddings for diffusion time."""

    def __init__(self, dim: int):
        super().__init__()
        self.dim = dim

    def forward(self, t: torch.Tensor) -> torch.Tensor:
        device = t.device
        half_dim = self.dim // 2
        embeddings = math.log(10000) / (half_dim - 1)
        embeddings = torch.exp(torch.arange(half_dim, device=device) * -embeddings)
        embeddings = t[:, None] * embeddings[None, :]
        return torch.cat([torch.sin(embeddings), torch.cos(embeddings)], dim=-1)


class ConvBlock(nn.Module):
    """Conv block with GroupNorm, SiLU, and time conditioning."""

    def __init__(self, in_ch: int, out_ch: int, time_dim: int):
        super().__init__()
        self.conv1 = nn.Conv2d(in_ch, out_ch, 3, padding=1)
        self.conv2 = nn.Conv2d(out_ch, out_ch, 3, padding=1)
        self.norm1 = nn.GroupNorm(min(8, out_ch), out_ch)
        self.norm2 = nn.GroupNorm(min(8, out_ch), out_ch)
        self.time_mlp = nn.Linear(time_dim, out_ch)
        self.act = nn.SiLU()

    def forward(self, x: torch.Tensor, t_emb: torch.Tensor) -> torch.Tensor:
        h = self.act(self.norm1(self.conv1(x)))
        h = h + self.time_mlp(t_emb)[:, :, None, None]
        h = self.act(self.norm2(self.conv2(h)))
        return h


class ScoreNetwork(nn.Module):
    """U-Net style score network for lattice systems."""

    def __init__(
        self,
        in_channels: int = 1,
        base_channels: int = 32,
        time_embed_dim: int = 64,
        num_blocks: int = 3
    ):
        super().__init__()
        self.time_embed = SinusoidalTimeEmbedding(time_embed_dim)
        self.time_mlp = nn.Sequential(
            nn.Linear(time_embed_dim, time_embed_dim * 2),
            nn.SiLU(),
            nn.Linear(time_embed_dim * 2, time_embed_dim)
        )

        # Build encoder, middle, decoder...
        # (Full implementation in docs/IMPLEMENTATION_PLAN.md)

    def forward(self, x: torch.Tensor, t: torch.Tensor) -> torch.Tensor:
        """Predict score s(x,t) ≈ ∇log p(x,t)."""
        # Implementation details in plan
        pass
```

```python
# backend/ml/models/diffusion.py
import torch

class DiffusionProcess:
    """VP-SDE diffusion process."""

    def __init__(self, beta_min: float = 0.1, beta_max: float = 20.0):
        self.beta_min = beta_min
        self.beta_max = beta_max

    def beta(self, t: torch.Tensor) -> torch.Tensor:
        """Linear beta schedule."""
        return self.beta_min + t * (self.beta_max - self.beta_min)

    def noise_level(self, t: torch.Tensor):
        """Compute alpha_t and sigma_t."""
        integral = self.beta_min * t + 0.5 * (self.beta_max - self.beta_min) * t**2
        alpha_t = torch.exp(-0.5 * integral)
        sigma_t = torch.sqrt(1 - alpha_t**2)
        return alpha_t, sigma_t

    def forward(self, x_0: torch.Tensor, t: torch.Tensor):
        """Sample x_t ~ q(x_t|x_0)."""
        alpha_t, sigma_t = self.noise_level(t)
        noise = torch.randn_like(x_0)
        x_t = alpha_t[:, None, None, None] * x_0 + sigma_t[:, None, None, None] * noise
        return x_t, noise
```

## Technical Challenges

1. **Time embedding** - Must condition score on diffusion time smoothly
2. **Architecture size** - Balance expressiveness vs CPU inference speed
3. **Numerical stability** - Score can be large at low noise levels (small t)

## Verification

```python
def test_score_network_output_shape():
    net = ScoreNetwork(in_channels=1, base_channels=32)
    x = torch.randn(4, 1, 32, 32)
    t = torch.rand(4)
    score = net(x, t)
    assert score.shape == x.shape

def test_time_embedding_dimension():
    embed = SinusoidalTimeEmbedding(64)
    t = torch.rand(8)
    out = embed(t)
    assert out.shape == (8, 64)

def test_diffusion_forward_variance():
    diff = DiffusionProcess()
    x_0 = torch.randn(100, 1, 16, 16)
    t = torch.ones(100) * 0.5
    x_t, noise = diff.forward(x_0, t)
    # Variance should be between x_0 variance and 1
    assert x_t.var() > x_0.var() * 0.3
```

## Definition of Done

- Score network compiles and forward pass works
- Output shape matches input shape
- Gradients flow through all parameters
- Memory usage reasonable (< 2GB for batch=16, size=64)

## Dependencies

- Phase 0: Foundation
- Phase 1: Ising Model (for testing with physical data)

## Related

Meta issue: #1
